clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -468,13 +522,37 @@ private void handleCompletedReceives(List<ClientResponse> responses, long now) {
             String source = receive.source();
             InFlightRequest req = inFlightRequests.completeNext(source);
             AbstractResponse body = parseResponse(receive.payload(), req.header);
+            log.trace("Completed receive from node {}, for key {}, received {}", req.destination, req.header.apiKey(), body);
             if (req.isInternalMetadataRequest)
                 metadataUpdater.handleCompletedMetadataResponse(req.header, now, body);
+            if (req.header.apiKey() == ApiKeys.API_VERSIONS.id)
+                handleApiVersionsResponse(req, now, body.toStruct());
             else
                 responses.add(req.completed(body, now));
         }
     }
 
+    private void handleApiVersionsResponse(InFlightRequest req, long now, Struct responseBody) {
+        ApiVersionsResponse apiVersionsResponse = new ApiVersionsResponse(responseBody);
+        final String node = req.destination;
+        for (ApiVersionsResponse.ApiVersion requiredApiVersion : requiredApiVersions) {
+            final ApiVersionsResponse.ApiVersion supportedApiVersion = apiVersionsResponse.apiVersion(requiredApiVersion.apiKey);
+            if (supportedApiVersion == null) {
+                close(node);
+                throw new KafkaException("Node " + req.destination + " does not support Api " + requiredApiVersion.apiKey + ".");
+            }
+            if (supportedApiVersion.maxVersion < requiredApiVersion.minVersion || supportedApiVersion.minVersion > requiredApiVersion.maxVersion) {
```

Comment:
Yeah, it's personal preference, so feel free to ignore. The `else if` makes the relation between the two cases stand out a bit more.